crates/goose/src/providers/azure.rs

```diff
@@ -0,0 +1,141 @@
+use anyhow::Result;
+use async_trait::async_trait;
+use reqwest::Client;
+use serde_json::Value;
+use std::time::Duration;
+
+use super::base::{ConfigKey, Provider, ProviderMetadata, ProviderUsage, Usage};
+use super::errors::ProviderError;
+use super::formats::openai::{create_request, get_usage, response_to_message};
+use super::utils::{emit_debug_trace, get_model, handle_response_openai_compat, ImageFormat};
+use crate::message::Message;
+use crate::model::ModelConfig;
+use mcp_core::tool::Tool;
+
+pub const AZURE_DEFAULT_MODEL: &str = "gpt-4o";
+pub const AZURE_DOC_URL: &str =
+    "https://learn.microsoft.com/en-us/azure/ai-services/openai/concepts/models";
+pub const AZURE_API_VERSION: &str = "2024-10-21";
+pub const AZURE_OPENAI_KNOWN_MODELS: &[&str] = &[
+    "gpt-4o",
+    "gpt-4o-mini",
+    "o1",
+    "o1-mini",
+    "o1-preview",
+    "gpt-4",
+];
+
+#[derive(Debug, serde::Serialize)]
+pub struct AzureProvider {
+    #[serde(skip)]
+    client: Client,
+    endpoint: String,
+    api_key: String,
+    deployment_name: String,
+    model: ModelConfig,
+}
+
+impl Default for AzureProvider {
+    fn default() -> Self {
+        let model = ModelConfig::new(AzureProvider::metadata().default_model);
+        AzureProvider::from_env(model).expect("Failed to initialize Azure OpenAI provider")
+    }
+}
+
+impl AzureProvider {
+    pub fn from_env(model: ModelConfig) -> Result<Self> {
+        let config = crate::config::Config::global();
+        let api_key: String = config.get_secret("AZURE_OPENAI_API_KEY")?;
+        let endpoint: String = config.get("AZURE_OPENAI_ENDPOINT")?;
+        let deployment_name: String = config.get("AZURE_OPENAI_DEPLOYMENT_NAME")?;
+
+        let client = Client::builder()
+            .timeout(Duration::from_secs(600))
+            .build()?;
+
+        Ok(Self {
+            client,
+            endpoint,
+            api_key,
+            deployment_name,
+            model,
+        })
+    }
+
+    async fn post(&self, payload: Value) -> Result<Value, ProviderError> {
+        let url = format!(
```

Comment:
can we use the Url crate to construct this - https://github.com/block/goose/pull/985